libs/libsamplerate/include/samplerate.hpp

```diff
@@ -0,0 +1,181 @@
+}
+
+namespace detail {
+    [[noreturn]] inline void throwError(int errorCode) { throw std::runtime_error(src_strerror(errorCode)); }
+    inline void throwOnError(int errorCode)
+    {
+        if(errorCode)
+            throw std::runtime_error(src_strerror(errorCode));
+    }
+
+    class State
+    {
+        SRC_STATE* state_;
+
+    public:
+        explicit State(SRC_STATE* state) : state_(state) {}
+        ~State() { src_delete(state_); }
+        State(const State& other) : state_(nullptr)
+        {
+            if(other.state_)
+            {
+                int error;
+                state_ = src_clone(other.state_, &error);
+                if(!state_)
+                    throwOnError(error);
+            }
+        }
+        State(State&& other) noexcept : state_(other.state_) { other.state_ = nullptr; }
+        friend void swap(State& lhs, State& rhs) { std::swap(lhs.state_, rhs.state_); }
+        State& operator=(State other)
+        {
+            swap(*this, other);
+            return *this;
+        }
+        SRC_STATE* get() { return state_; }
+        operator SRC_STATE*() { return state_; }
+    };
+
+    class StateBase
+    {
+    protected:
+        State state_;
+        explicit StateBase(SRC_STATE* state) : state_(state) {}
+
+    public:
+        void reset() { throwOnError(src_reset(state_)); }
+        void setRatio(double newRatio) { throwOnError(src_set_ratio(state_, newRatio)); }
+        int getChannels() { return src_get_channels(state_); }
+        /// Get internal state
+        SRC_STATE* getState() { return state_; }
+    };
+} // namespace detail
+
+struct Data
+{
+    SRC_DATA data;
+    Data(const float* in, size_t input_frames, float* out, size_t output_frames, double src_ratio, bool end_of_input = true) : data{}
+    {
+        data.data_in = in;
+        data.data_out = out;
+        data.input_frames = static_cast<long>(input_frames);
+        data.output_frames = static_cast<long>(output_frames);
+        data.src_ratio = src_ratio;
+        data.end_of_input = end_of_input;
+    }
+    operator SRC_DATA*() { return &data; }
+    long input_frames_used() const { return data.input_frames_used; }
+    long output_frames_gen() const { return data.output_frames_gen; }
+};
+
+struct Result
+{
+    long input_frames_used, output_frames_gen;
+};
+
+inline Result simple(Data& data, Converter converter, int channels)
+{
+    detail::throwOnError(src_simple(&data.data, static_cast<int>(converter), channels));
+    return {data.input_frames_used(), data.output_frames_gen()};
+}
+
+inline Result simple(Data&& data, Converter converter, int channels)
+{
+    return simple(static_cast<Data&>(data), converter, channels);
+}
+
+class State : public detail::StateBase
+{
+    static SRC_STATE* createOrThrow(Converter converter, int channels)
+    {
+        int error;
+        auto* state = src_new(static_cast<int>(converter), channels, &error);
+        if(!state)
```

Comment:
shouldn't you use `throwOnError` here? instead of the if

Reply:
Yes and no. I want to throw an error if state is NULL, which *SHOULD* be the case if `error!=0`, but to be safe I'll also throw for `state=NULL && error=0`